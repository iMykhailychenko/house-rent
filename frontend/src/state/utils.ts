import { ErrorState } from './interfaces/common';

export const formatSeverError = (error: Error & { status?: number; message?: string }): ErrorState => ({
    status: error.status,
    message: error.message || 'something_went_wrong',
});
