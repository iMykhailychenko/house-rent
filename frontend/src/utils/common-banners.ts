import { Banner, BannerType } from '../state/entities/banners/banners.interface';

export const POST_IN_ARCHIVE: Banner = {
    id: 'POST_IN_ARCHIVE',
    type: BannerType.ERROR,
    content:
        'УВАГА! Цей пост знаходиться в архіві. Щоб інші користувачі могли його побачити вам потрібно змінити статус цього посту на активний!',
};

export const POST_IN_DRAFT: Banner = {
    id: 'POST_IN_DRAFT',
    type: BannerType.INFO,
    content:
        'УВАГА! Цей пост не активований. Щоб інші користувачі могли його побачити вам потрібно змінити статус цього посту на активний!',
};

export const VALIDATE_EMAIL_WARN: Banner = {
    id: 'VALIDATE_EMAIL_WARN',
    type: BannerType.WARNING,
    content:
        'Ви не активували вашу електронну пошту. Щоб повною мірою використовувати функціонал сайту перейдіть за посиланням в листі',
};

export const VALIDATE_EMAIL_ERROR: Banner = {
    id: 'VALIDATE_EMAIL_ERROR',
    type: BannerType.ERROR,
    content:
        'Ви не активували вашу електронну пошту. Щоб повною мірою використовувати функціонал сайту перейдіть за посиланням в листі',
};
