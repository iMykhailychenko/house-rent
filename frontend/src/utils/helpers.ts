import cookie from 'cookie';

interface ParseCookieParams<T> {
    value?: string;
    key: string;
    isJson?: boolean;
    defaultValue: T;
}

export const parseCookie = <T>({ value = '', key, isJson = false, defaultValue }: ParseCookieParams<T>): T => {
    try {
        return isJson ? cookie.parse(value)[key] || defaultValue : JSON.parse(cookie.parse(value)[key]);
    } catch (error) {
        return defaultValue;
    }
};
