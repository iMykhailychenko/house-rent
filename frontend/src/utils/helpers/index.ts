export { copyText } from './copy-text.helper';
export { addMonthToDate, formatDate, formatTime, validateDate, onlineStatus } from './date.helper';
export { cutString } from './string.helper';
export { parseCookie } from './cookie.helper';
export { addZero } from './number.helper';
export { shallowRouter } from './shallow-router.helper';
export { errorNotif } from './error-logger.helper';
