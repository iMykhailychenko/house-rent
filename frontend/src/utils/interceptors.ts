import axios from 'axios';
import * as AxiosLogger from 'axios-logger';

const api = axios.create({});

api.interceptors.request.use(
    req => AxiosLogger.requestLogger(req, { data: false }),
    error => Promise.reject(error),
);

api.interceptors.response.use(
    res => AxiosLogger.responseLogger(res, { data: false }),
    error => {
        if (error?.response?.status === 401) {
            // delete api.defaults.headers.common.Authorization;
        }

        return Promise.reject(error);
    },
);

export default api;
