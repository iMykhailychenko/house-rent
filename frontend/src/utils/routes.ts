const routes = {
    home: '/',
    private: '/private',
    chat: '/chat',
    auth: {
        login: '/auth/login',
        join: '/auth/join',
        success: '/auth/success',
        reset: '/auth/reset',
    },
    users: {
        profile: (id: string | number): string => `/users/${id}`,
    },
    posts: {
        list: '/posts',
        new: '/posts/new',
        single: (id: string | number): string => `/posts/${id}`,
    },
};

export default routes;
