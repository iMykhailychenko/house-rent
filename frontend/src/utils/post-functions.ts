import { NextRouter } from 'next/router';
import { toast } from 'react-toastify';

import { modal } from '../components/common/modal/modal';
import toastConfig from '../config/toast.cofig';
import { ThunkAppDispatch } from '../hooks/redux.hook';
import { POST_STATUS } from '../state/entities/posts/posts.interface';
import { personalPostsListThunk } from '../state/entities/posts/thunks/personal-posts.thunk';
import { updatePostStatusThunk } from '../state/entities/posts/thunks/update-status.thunk';

import routes from './routes';

export enum POST_ACTIONS {
    ARCHIVE = 'archive',
    EDIT = 'edit',
    ACTIVATE = 'activate',
    DELETE = 'delete',
}

type PostActionsMap = { [key in POST_STATUS]: POST_ACTIONS[] };
export const postActionsMap: PostActionsMap = {
    [POST_STATUS.DRAFT]: [POST_ACTIONS.ACTIVATE, POST_ACTIONS.EDIT, POST_ACTIONS.ARCHIVE, POST_ACTIONS.DELETE],
    [POST_STATUS.ACTIVE]: [POST_ACTIONS.ARCHIVE, POST_ACTIONS.EDIT, POST_ACTIONS.DELETE],
    [POST_STATUS.ARCHIVE]: [POST_ACTIONS.ACTIVATE, POST_ACTIONS.DELETE, POST_ACTIONS.EDIT],
};

type PostFunctionsMap = { [key in POST_ACTIONS]: (postId: number) => Promise<void> };
export const postFunctionsMap = (dispatch: ThunkAppDispatch, history: NextRouter): PostFunctionsMap => {
    const page = +String(history.query.page || 1);
    const status = String(history.query.status || POST_STATUS.DRAFT) as POST_STATUS;

    return {
        [POST_ACTIONS.ARCHIVE]: async (postId: number): Promise<void> => {
            modal.close();
            const post = await dispatch(updatePostStatusThunk({ id: postId, status: POST_STATUS.ARCHIVE })).unwrap();
            if (!post.id) return;
            await dispatch(personalPostsListThunk({ status, page }));
            toast.success('Ви успішно перемістили пост до Архіву. Тепер він прихований від інших користувачів', toastConfig);
        },
        [POST_ACTIONS.EDIT]: async (postId: number): Promise<void> => {
            modal.close();
            history.push(routes.posts.edit(postId));
        },
        [POST_ACTIONS.ACTIVATE]: async (postId: number): Promise<void> => {
            modal.close();
            const post = await dispatch(updatePostStatusThunk({ id: postId, status: POST_STATUS.ACTIVE })).unwrap();
            if (!post.id) return;
            await dispatch(personalPostsListThunk({ status, page }));
            toast.success('Ви успішно активували свій пост! Тепер він доступний для інших користувачів на сайті', toastConfig);
        },
        [POST_ACTIONS.DELETE]: async (postId: number): Promise<void> => {
            modal.close();
            const post = await dispatch(updatePostStatusThunk({ id: postId, status: POST_STATUS.ACTIVE })).unwrap();
            if (!post.id) return;
            await dispatch(personalPostsListThunk({ status, page }));
            toast.success('Ви успішно активували свій пост! Тепер він доступний для інших користувачів на сайті', toastConfig);
        },
    };
};
