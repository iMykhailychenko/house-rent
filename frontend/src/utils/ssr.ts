import { Store } from '@reduxjs/toolkit';
import axios from 'axios';
import { GetServerSideProps, GetServerSidePropsContext, Redirect } from 'next';

import { cookiesTokenAction } from '../state/entities/auth/auth.reducer';
import { profileInfoThunk } from '../state/entities/profile/profile.thunk';
import rootInitialState from '../state/initial-state';
import { RootState } from '../state/reducer';
import { initializeStore } from '../state/store';

import routes from './routes';

type ReturnCallback<T> = { props: T | { state: RootState } };
type SSRCallback<T> = (
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    context: GetServerSidePropsContext & { store?: Store<RootState, any> },
) => Promise<ReturnCallback<T> | void>;

export const withStore = <T>(callback?: SSRCallback<T>): GetServerSideProps => {
    return async (context: GetServerSidePropsContext): Promise<ReturnCallback<T>> => {
        const store = initializeStore(rootInitialState);

        const token = axios.defaults.headers.common.Authorization;
        if (token) {
            await store.dispatch(profileInfoThunk());
            await store.dispatch(cookiesTokenAction(token));
        }

        if (callback) {
            const result = await callback({ ...context, store });
            return { props: { ...(result?.props || {}), state: store.getState() } };
        }

        return { props: { state: store.getState() } };
    };
};

export const withAuthRedirect =
    <T>(callback: SSRCallback<T> | null, reverse = false): GetServerSideProps =>
    async (context: GetServerSidePropsContext): Promise<ReturnCallback<T> | { redirect: Redirect }> => {
        const token = axios.defaults.headers.common.Authorization;
        if (reverse ? token : !token) {
            return {
                redirect: {
                    statusCode: 302,
                    destination: routes.auth.login,
                },
            };
        }

        const store = initializeStore(rootInitialState);

        if (callback) {
            const result = await callback({ ...context, store });
            return { props: { ...(result?.props || {}), state: store.getState() } };
        }

        return { props: { state: store.getState() } };
    };
