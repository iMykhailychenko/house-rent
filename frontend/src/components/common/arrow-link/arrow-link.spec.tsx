import { render } from '../../../tests/tests.utils';
import ArrowLink from './arrow-link';

describe('Home page', () => {
    it('matches snapshot', () => {
        const { asFragment } = render(<ArrowLink href="test">test</ArrowLink>);
        expect(asFragment()).toMatchSnapshot();
    });

    it('matches snapshot with props', () => {
        const { asFragment } = render(
            <ArrowLink href="test" className="test" direction="right">
                test
            </ArrowLink>,
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
