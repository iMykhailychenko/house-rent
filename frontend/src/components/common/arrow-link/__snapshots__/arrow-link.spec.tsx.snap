// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page matches snapshot 1`] = `
<DocumentFragment>
  <header
    class="header"
  >
    <div
      class="inner md root"
    >
      <div
        class="item"
      >
        <button
          class="button menu"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
            />
          </svg>
        </button>
        <div
          class="root"
        >
          <button
            class="button active"
            type="button"
          >
            ua
          </button>
          <button
            class="button"
            type="button"
          >
            ru
          </button>
        </div>
      </div>
      <div
        class="item"
      >
        <button
          class="root"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69zM12 18c-3.31 0-6-2.69-6-6s2.69-6 6-6 6 2.69 6 6-2.69 6-6 6zm0-10c-2.21 0-4 1.79-4 4s1.79 4 4 4 4-1.79 4-4-1.79-4-4-4z"
            />
          </svg>
        </button>
        <div
          class="root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
            />
          </svg>
          <button
            class="button"
            type="button"
          >
            увійти
          </button>
          <button
            class="button"
            type="button"
          >
            зареєструватися
          </button>
        </div>
      </div>
    </div>
  </header>
  <a
    class="link"
    href="/test"
  >
    <span>
      test
    </span>
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
      />
    </svg>
  </a>
  <footer
    class="footer"
  >
    <div
      class="lg root"
    >
      Lorem ipsum dolor sit amet, consectetur adipisicing elit. Ad eligendi error est fugiat iure maxime molestiae nobis odit, pariatur reiciendis rem sit, unde velit? Earum labore magnam quas quos voluptas!
    </div>
  </footer>
</DocumentFragment>
`;

exports[`Home page matches snapshot with props 1`] = `
<DocumentFragment>
  <header
    class="header"
  >
    <div
      class="inner md root"
    >
      <div
        class="item"
      >
        <button
          class="button menu"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
            />
          </svg>
        </button>
        <div
          class="root"
        >
          <button
            class="button active"
            type="button"
          >
            ua
          </button>
          <button
            class="button"
            type="button"
          >
            ru
          </button>
        </div>
      </div>
      <div
        class="item"
      >
        <button
          class="root"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69zM12 18c-3.31 0-6-2.69-6-6s2.69-6 6-6 6 2.69 6 6-2.69 6-6 6zm0-10c-2.21 0-4 1.79-4 4s1.79 4 4 4 4-1.79 4-4-1.79-4-4-4z"
            />
          </svg>
        </button>
        <div
          class="root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
            />
          </svg>
          <button
            class="button"
            type="button"
          >
            увійти
          </button>
          <button
            class="button"
            type="button"
          >
            зареєструватися
          </button>
        </div>
      </div>
    </div>
  </header>
  <a
    class="link test"
    href="/test"
  >
    <span>
      test
    </span>
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
      />
    </svg>
  </a>
  <footer
    class="footer"
  >
    <div
      class="lg root"
    >
      Lorem ipsum dolor sit amet, consectetur adipisicing elit. Ad eligendi error est fugiat iure maxime molestiae nobis odit, pariatur reiciendis rem sit, unde velit? Earum labore magnam quas quos voluptas!
    </div>
  </footer>
</DocumentFragment>
`;
